service/matching/taskQueueManager.go

```diff
@@ -729,22 +742,38 @@ func (c *taskQueueManagerImpl) fetchMetadataFromRootPartition(ctx context.Contex
 		TaskQueue:                 rootTqName,
 		WantVersioningDataCurhash: curHash,
 	})
-	if err != nil {
-		return nil, err
-	}
 	// If the root partition returns nil here, then that means our data matched, and we don't need to update.
 	// If it's nil because it never existed, then we'd never have any data.
 	// It can't be nil due to removing versions, as that would result in a non-nil container with
 	// nil inner fields.
-	if res.VersioningData != nil {
-		c.db.setVersioningDataForNonRootPartition(res.VersioningData)
-		// Only start the poller if there's actually data to poll for
-		c.metadataPoller.StartIfUnstarted(func() { _, _ = c.fetchMetadataFromRootPartition(context.TODO()) })
+	if !res.GetMatchedReqHash() {
+		c.db.setVersioningDataForNonRootPartition(res.GetVersioningData())
 	}
-	return res.VersioningData, nil
+	// We want to start the poller as long as the root partition has any kind of data (or fetching hasn't worked)
+	if res.GetMatchedReqHash() || res.GetVersioningData() != nil || err != nil {
+		c.startMetadataPollerIfUnstarted()
+	}
+	if err != nil {
+		return nil, err
+	}
+	return res.GetVersioningData(), nil
+}
+
+func (c *taskQueueManagerImpl) startMetadataPollerIfUnstarted() {
+	c.metadataPoller.StartIfUnstarted(func() bool {
+		dat, err := c.fetchMetadataFromRootPartition(context.TODO())
+		if dat == nil && err == nil {
+			// Can stop polling since there is no versioning data. Loop will be restarted if we
+			// are told to invalidate the data, or we attempt to fetch it via GetVersioningData.
+			return true
+		}
+		return false
+	})
 }
 
-func (mp *metadataPoller) StartIfUnstarted(pollFn func()) {
+// StartIfUnstarted starts the poller if it's not already started. The passed in function is called repeatedly
+// and if it returns true, the poller will shut down, at which point it may be started again.
+func (mp *metadataPoller) StartIfUnstarted(pollFn func() bool) {
 	mp.launched.Do(func() {
 		go func() {
```

Comment:
and I would also say this anonymous func can be a named method on metadataPoller

Reply:
Done